providers/class.two-factor-backup-codes.php

```diff
@@ -0,0 +1,274 @@
+			return true;
+		}
+		return false;
+	}
+
+	/**
+	 * Returns the name of the provider.
+	 *
+	 * @since 0.1-dev
+	 */
+	public function get_label() {
+		return _x( 'Backup Verification Codes (Single Use)', 'Provider Label' );
+	}
+
+	/**
+	 * Whether this Two Factor provider is configured and available for the user specified.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 * @return boolean
+	 */
+	public function is_available_for_user( $user ) {
+		return true;
+	}
+
+	/**
+	 * Inserts markup at the end of the user profile field for this provider.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param WP_User $user WP_User object of the logged-in user.
+	 */
+	public function user_options( $user ) {
+		$user_id = $user->ID;
+		$backup_codes = get_user_meta( $user_id, self::BACKUP_CODES_META_KEY, true );
+		$ajax_nonce = wp_create_nonce( 'two-factor-backup-codes-generate-json' );
+		?>
+		<p id="two-factor-backup-codes">
+			<button type="button" class="button button-two-factor-backup-codes-generate button-secondary hide-if-no-js">
+				<?php esc_html_e( 'Generate Verification Codes' ); ?>
+			</button>
+			<?php echo wp_kses( sprintf( _n( '%s unused code remaining.', '%s unused codes remaining.', count( $backup_codes ) ), '<span class="two-factor-backup-codes-count">' . count( $backup_codes ) . '</span>' ), array( 'span' => array( 'class' => array() ) ) ); ?>
+		</p>
+		<div class="two-factor-backup-codes-wrapper" style="display:none;">
+			<ol class="two-factor-backup-codes-unused-codes"></ol>
+			<p class="description"><?php esc_html_e( "Write 'em down y'all!" ); ?></p>
+		</div>
+		<script type="text/javascript">
+			// @todo: move this into a JS file & internationalize the count strings
+			jQuery( document ).ready( function( $ ) {
+				$( '.button-two-factor-backup-codes-generate' ).click( function() {
+					jQuery.ajax( {
+						url: ajaxurl,
+						data: {
+							action: 'two_factor_backup_codes_generate',
+							nonce: '<?php echo esc_js( $ajax_nonce ); ?>'
+						},
+						dataType: 'JSON',
+						success: function( data ) {
+							$( '.two-factor-backup-codes-wrapper' ).show();
+							$( '.two-factor-backup-codes-unused-codes' ).html( '' );
+							$.each( data, function( key, val ) {
+								$( '.two-factor-backup-codes-unused-codes' ).append( '<li>' + val + '</li>' );
+							} );
+							// Update counter.
+							$( '.two-factor-backup-codes-count' ).html( data.length );
+						},
+						error: function( errorThrown ) {
+							alert( 'error' ); // @todo: remove for production
+							console.log( errorThrown );
+						}
+					} );
+				} );
+			} );
+		</script>
+		<?php
+	}
+
+	/**
+	 * Generates backup codes & updates the user meta.
+	 *
+	 * @since 0.1-dev
+	 *
+	 * @param int $user_id The logged-in user ID.
+	 */
+	public function generate_codes( $user_id ) {
+		$codes = array();
+		$codes_hashed = array();
+
+		for ( $i = 0; $i < self::NUMBER_OF_CODES; $i++ ) {
```

Comment:
I almost wonder if it may be tidier to do an `array_fill()` and then `array_map()` to populate the array instead of a for loop.  Probably fine to leave as is.